### PR TITLE
Milestone 3 - Users & posts

### PR DESCRIPTION
**Pull request by macnick**
Here is what we implemented in this milestone

1. Create User Model
2. Create Post Model
3. Install Devise gem for user authentication
4. Log In and Sign Up pages for users to create an account and log in
5. Automatic log in after account creation
6. User page where he/she can create posts and view all the his/hers previous posts
7. Capability to delete a post.
8. Unit tests for User and Post models
9. Integration tests for user log in and sign up

Please review our code.

### CODE REVIEW

**Review by Torres-ssf — CHANGES_REQUESTED** [review]
### Changes required :smile: 
Hi there team :hand: Good job so far, but we have changes to make before moving on.

- [x] I notice that you added the linter files to your repo but we also need to give it the authorization for checking your repo files. Check out [this](https://github.com/microverseinc/linters-config/tree/master/ruby) link.
Screenshot from 2020-01-06 21-14-16

- [x] The user's index controller needs to show all registered users. You don't need to create any buttons to add friends right now. Just make sure that the index method loads a view that shows all registered users' names. This will be improved in future milestones.
Screenshot from 2020-01-08 10-04-44

- [x] Good job implement the create post functionality :+1: . But as it said in the Odin project instructions, we need to display created posts with content and author.
Screenshot from 2020-01-08 09-59-49

- [x] Right now, one of your tests is falling
Screenshot from 2020-01-08 10-20-01

Please submit for code review again after making these changes.
Feel free to message me on slack if you have any questions
Happy coding :computer:

**Comment by macnick** [comment]
Hello, 
Thank you for the prompt review. It was very helpful.

We fixed the following: 
1. Gave authorization to stickler to check our repo. It created a pull request out of Milestone-3 and found no errors. We will wait for your ok to merge it with master.
2. Changed the  index controller to show all registered users with a fake button "Add friend" 
3. Modified posts index to include author for each post and of course the post text.
4. We forgot to write that for the first test to work you have to create a user with email "nick@test.com" and password "123456" first. The test, tests that an registered user can login. In a new database there is no registered user. As you see in the second test we create random users. 

Please review our changes.

File: .rubocop.yml [hunk 1]
```diff
@@ -0,0 +1,34 @@
+AllCops:
+  Exclude:
+    - "bin/*"
+    - "db/*"
+    - "config/*"
+    - "Guardfile"
+    - "Rakefile"
+  DisplayCopNames: true
+
+Metrics/LineLength:
```
Comment (mosaaleb) [inline .rubocop.yml]: - [x] There is no `Metrics/LineLength` rule. It's `Layout/LineLength`. Check it [here](https://rubocop.readthedocs.io/en/latest/cops_layout/#layoutlinelength)

File: app/controllers/posts_controller.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
+# frozen_string_literal: true
+
+class PostsController < ApplicationController
+  def index
+    @post = Post.new
+    @user_posts = Post.all #current_user.posts
+  end
+
+  def create
+    @user = current_user
```
Comment (mosaaleb) [inline app/controllers/posts_controller.rb]: - [x] **[OPTIONAL]** `current_user` is already available in our views. We can remove this line and use `current_user` for the post creation.

File: app/controllers/posts_controller.rb [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+  def destroy
+    @post = Post.find(params[:id])
+    @post.destroy
+    redirect_to posts_path
+  end
+
+  private
+
+  def post_params
+    params.require(:post).permit(:post_text, :user_id)
```
Comment (mosaaleb) [inline app/controllers/posts_controller.rb]: - [x] **[OPTOINAL]** We don't need to permit the `user_id`. We already creating the post through association with `current_user`

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
+
+class UsersController < ApplicationController
+  def index
+    @users = User.all
+  end
+
+  def show
+    @user = User.find(current_user.id)
```
Comment (mosaaleb) [inline app/controllers/users_controller.rb:9]: - [x] All show pages for all users are currently only displaying the current user. -- we need find the user from the id params `users/1`

**Review by mosaaleb — CHANGES_REQUESTED** [review]
### Well done @macnick @addod19 :clap:


- [x] Nice job adding stickler, but there are still stickler errors. Let's fix them. See my comment for this error.

- [x] I created a post with a user, and was able to delete the post with another user. -- We can display the delete button in our views only if the current user is the author if the post. -- Don't forget to wrap this logic (checking whether the current user is the author) in a helper method. [helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)

- [x] Failed test: 
failed-test
The problem with this test is that you don't have a user with these credentials yet. 

I suggest that you go through the first 4 lessons in this course [TDD](https://thoughtbot.com/upcase/test-driven-rails)

Also we don't have integration test for adding posts. -- let's add it.

#### Please handle these requirements and submit a new code review.

##### Muhammad Ebeid

File: app/controllers/posts_controller.rb (hunk 2 above)
Comment (addod19) [reply app/controllers/posts_controller.rb]: removed line 10 and set @post = current_user.posts.build(post_params) as required by reviewer.

File: app/controllers/posts_controller.rb (hunk 3 above)
Comment (addod19) [reply app/controllers/posts_controller.rb]: removed user_id from post_params

File: app/controllers/users_controller.rb (hunk 4 above)
Comment (addod19) [reply app/controllers/users_controller.rb:9]: Can you please create two users and two posts to see all the posts. Attached is the screenshot of the post index page
Screen Shot 2020-01-09 at 11 02 04

**Comment by macnick** [comment]
Hello,
Thank you for the prompt review. It was very helpful.

We fixed the following:

1. Fixed stickler to work properly.
2. Fixed the delete button to appear only to the current user's posts.
3. Changed the logic for the test so first it creates the users and then test for login
4. Added integration test for creating post.
5. Also fixed the optional requirements.

Please review our changes.

**Review by mosaaleb — APPROVED** [review]
### Well done Team :clap:

#### All requirements are handled. All tests have passed.

#### You can merge :+1: 

##### Muhammad Ebeid

**Comment by addod19** [comment]
Thank you for helping us.

**Pull request merged**